File: wk4/java/threeSum.java
// problem link: https://leetcode.com/problems/3sum/description/

class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        ArrayList<List<Integer>> ret = new ArrayList<>();
        for(int i = 0; i < nums.length - 2; i++) {
            if (i != 0 && nums[i] == nums[i-1])
                continue;
            else {
                int j = i + 1, k = nums.length - 1;
                while(j < k) {
                    int sum = nums[i] + nums[j] + nums[k];
                    if(sum > 0)
                        k--;
                    else if (sum < 0)
                        j++;
                    else {
                        ArrayList<Integer> temp = new ArrayList<>();
                        temp.add(nums[i]);
                        temp.add(nums[j]);
                        temp.add(nums[k]);
                        ret.add(temp);
                        j++;
                        while(j < nums.length && nums[j] == nums[j-1])
                            j++;
                    }
                }
            }
        }
        return ret;
    }
}

File: wk4/cpp/threesum.cpp
// problem link: https://leetcode.com/problems/3sum/description/

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(begin(nums), end(nums));
        vector<vector<int>> ret;
        for(int i = 0; i < nums.size() - 2; i++) {
            if(i && nums[i] == nums[i - 1])
                continue;
            int j = i + 1, k = nums.size() - 1;
            while(j < k) {
                int sum = nums[i] + nums[j] + nums[k];
                if(sum > 0)
                    k--;
                else if(sum < 0)
                    j++;
                else {
                    vector<int> temp = {nums[i], nums[j], nums[k]};
                    ret.push_back(temp);
                    j++;
                    while(j < nums.size() && nums[j] == nums[j - 1])
                        j++;
                }
            }
        }
        return ret;
    }
};